examples/step-81/doc/results.dox

```diff
@@ -0,0 +1,296 @@
+essentially an implementation of $e^{-i\omega t}(Re(E) + i*Im(E))$ as we
+increment time. This is done by slightly changing the output function to
+generate a series of .vtk files, which will represent out solution wave as
+we increment time. Introduce an input variable $t$ in the output_results()
+class as output_results(unsigned int t). Then change the class itself to
+the following:
+
+@code
+template <int dim>
+void Maxwell<dim>::output_results(unsigned int t)
+{
+  std::cout << "Running step:" << t << std::endl;
+  DataOut<2> data_out;
+  data_out.attach_dof_handler(dof_handler);
+  Vector<double> postprocessed;
+  postprocessed.reinit(solution);
+  for (unsigned int i = 0; i < dof_handler.n_dofs(); ++i)
+    {
+      if (i % 4 == 0)
+        {
+          postprocessed[i] = std::cos(2 * M_PI * 0.04 * t) * solution[i] -
+                             std::sin(2 * M_PI * 0.04 * t) * solution[i + 1];
+        }
+      else if (i % 4 == 2)
+        {
+          postprocessed[i] = std::cos(2 * M_PI * 0.04 * t) * solution[i] -
+                             std::sin(2 * M_PI * 0.04 * t) * solution[i + 1];
+        }
+    }
+  data_out.add_data_vector(postprocessed, {"E_x", "E_y", "null0", "null1"});
+  data_out.build_patches();
+  const std::string filename =
+    "solution-" + Utilities::int_to_string(t) + ".vtk";
+  std::ofstream output(filename);
+  data_out.write_vtk(output);
+  std::cout << "Done running step:" << t << std::endl;
+}
+@endcode
+
+Finally, in the run() function, replace output_results() with
+@code
+for (int t = 0; t <= 100; t++)
+  {
+    output_results(t);
+  }
+@endcode
+
+This would generate 100 solution .vtk files, which can be opened in a group
+on Paraview and then can be saved as an animation. We used FFMPEG to
+generate gifs.
+
+<h3> Possibilities for Extension </h3>
+
+The example step could be extended in a number of different directions.
+<ul>
+  <li>
+    The current program uses a direct solver to solve the linear system.
+    This is efficient for two spatial dimensions where scattering problems
+    up to a few millions degrees of freedom can be solved. In 3D, however,
+    the increased stencil size of the Nedelec element pose a severe
+    limiting factor on the problem size that can be computed. As an
+    alternative, the idea to use iterative solvers can be entertained.
+    This, however requires specialized preconditioners. For example, just
+    using an iterative Krylov space solver (such as SolverGMRES) on above
+    problem will requires many thousands of iterations to converge.
+    Unfortunately, time-harmonic Maxwell's equations lack the usual notion
+    of local smoothing properties, which renders the usual suspects, such
+    as a geometric multigrid (see the Multigrid class), largely useless. A
+    possible extension would be to implement a Schwarz preconditioner
+    (based on domain decomposition), or a sweeping preconditioner.
```

Comment:
Isn't this the context of edge-based multigrid with particular structure-preserving smootheners? That topic was pioneered by Hiptmair if I remember correctly, and there have been recent contributions, e.g., https://onlinelibrary.wiley.com/doi/abs/10.1002/nla.2348 ? I am not sufficiently well into the H(curl) and Maxwell context to say whether the equations are the same, but I guess that we should not say that multigrid is largely useless, when that statement only applies to the naive smoothers we use for Poisson, elasticity or Stokes?

Reply:
If I remember correctly, the work of Hiptmair is related to the "well-posed" Maxwell equation, i.e.:
curl( mu^(-1) curl( E ) ) **+** epsilon E = 0

This is quite complicated,
Standard Multigrid methods will fail (e.g.: see: [Geometric multigrid with multiplicative Schwarz smoothers for eddy-current and Maxwell’s equations in deal.II](https://doi.org/10.1016/j.exco.2021.100027) or [Why it is difficult to solve Helmholtz problems with classical iterative methods](https://link.springer.com/chapter/10.1007/978-3-642-22061-6_10#:~:text=Simply%20using%20a%20Krylov%20method,do%20not%20remedy%20the%20situation))
Even a more advanced Multigrid preconditioner, would have the problem, that especially for higher wavenumber we can not resolve the waves sufficiently with a coarse grid. So we still need a preconditioner, that can handle these higher frequencies (e.g. a preconditioner based on domain decomposition methods).

Reply:
One further remark, but this is just a minor detail:
I am not sure, that a preconditioner based on a _standart Schwarz Domain Decomposition_ will work for the time-harmonic Maxwell equations.
A preconditioner based on an _optimised Schwarz Domain Decomposition_  should be much more suitable for the problem.

Reply:
I see, the problem here is the indefinite Maxwell problem, I always confuse the sign in the Maxwell equations. Thank you for the explanation. With that info, I agree with the basic description and problems. Either way, this is only a small outlook into what topics the current state-of-the-art is considering, so the current state is fine.

Reply:
@kinnewig Good point! We will add two references specifying what preconditioners we have in mind precisely.